Main.hs

```diff
@@ -2,32 +2,142 @@ module Main where
 
 import Control.Monad (unless)
 import Text.Printf (printf)
+import Data.Either
 
-data Expr
+data Expr =
+  Num Double
+  | Sqrt Expr
+  | Add Expr Expr
+  | Sub Expr Expr
+  | Mul Expr Expr
+  | Div Expr Expr
+  | Pow Expr Expr
 
 instance Show Expr where
-  show = undefined
+  show (Num a) = show a
+  show (Sqrt a) = show "sqrt(" ++ show a ++ ")" 
+  show (Add x y) = "(" ++ show x ++ " + " ++ show y ++ ")"
+  show (Sub x y) = "(" ++ show x ++ " - " ++ show y ++ ")"
+  show (Mul x y) = "(" ++ show x ++ " * " ++ show y ++ ")"
+  show (Div x y) = "(" ++ show x ++ " / " ++ show y ++ ")"
+  show (Pow x y) = "(" ++ show x ++ " ^ " ++ show y ++ ")"
 
 instance Eq Expr where
-  (==) = undefined
+  (==) (Num a) (Num b) = a == b
+  (==) (Sqrt a) (Sqrt b) = a == b
+  (==) (Add a b) (Add c d) = (a == c) && (b == d)
+  (==) (Sub a b) (Sub c d) = (a == c) && (b == d)
+  (==) (Mul a b) (Mul c d) = (a == c) && (b == d)
+  (==) (Div a b) (Div c d) = (a == c) && (b == d)
+  (==) (Pow a b) (Pow c d) = (a == c) && (b == d)
+  (==) _ _ = False
 
-data Error
+
+data Error = DivZero Expr | NegRoot Expr
 
 instance Show Error where
-  show = undefined
+  show (DivZero expr) = "division by zero is forbidden " ++ show expr
+  show (NegRoot expr) = "sqrt of a negative number is forbidden " ++ show expr
 
 instance Eq Error where
-  (==) = undefined
+  (==) x y = show x == show y
+
+evenRoots = [1/x | x<-[2,4..31]]  -- we can take more here but I think 1e-9 is pretty small enough
+
+performOperation :: (Double -> Double -> Double) -> Expr -> Expr -> Either Error Double
+performOperation func a b
+  | isLeft x = x
+  | isLeft y = y
+  | otherwise = Right (func (fromRight 1 x) (fromRight 1 y))
```

Comment:
Using `fromRight` is awkward: you have to use `1` here, which doesn't make a lot of sense. If the line above is modified or accidentally deleted, you'll have this `1` appearing at run time out of nowhere. It's better to use pattern matching so that an error is reported at compile time instead.

Reply:
This comment is also applicable to the code below.